src/main/java/org/opensearch/neuralsearch/processor/NeuralSparseTwoPhaseProcessor.java

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.opensearch.neuralsearch.processor;
+
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.Multimap;
+import lombok.Getter;
+import lombok.Setter;
+import org.opensearch.common.SetOnce;
+import org.opensearch.action.search.SearchRequest;
+import org.opensearch.index.query.BoolQueryBuilder;
+import org.opensearch.index.query.QueryBuilder;
+import org.opensearch.ingest.ConfigurationUtils;
+import org.opensearch.neuralsearch.query.NeuralSparseQueryBuilder;
+import org.opensearch.search.builder.SearchSourceBuilder;
+import org.opensearch.search.pipeline.AbstractProcessor;
+import org.opensearch.search.pipeline.Processor;
+import org.opensearch.search.pipeline.SearchRequestProcessor;
+import org.opensearch.search.rescore.QueryRescorerBuilder;
+import org.opensearch.search.rescore.RescorerBuilder;
+
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * A SearchRequestProcessor to generate two-phase NeuralSparseQueryBuilder,
+ * and add it to the Rescore of a searchRequest.
+ */
+@Setter
+@Getter
+public class NeuralSparseTwoPhaseProcessor extends AbstractProcessor implements SearchRequestProcessor {
+
+    public static final String TYPE = "neural_sparse_two_phase_processor";
+    private boolean enabled;
+    private float ratio;
+    private float window_expansion;
+    private int max_window_size;
+    private static final String PARAMETER_KEY = "two_phase_parameter";
+    private static final String RATIO_KEY = "prune_ratio";
+    private static final String ENABLE_KEY = "enabled";
+    private static final String EXPANSION_KEY = "expansion_rate";
+    private static final String MAX_WINDOW_SIZE_KEY = "max_window_size";
+
+    protected NeuralSparseTwoPhaseProcessor(
+        String tag,
+        String description,
+        boolean ignoreFailure,
+        boolean enabled,
+        float ratio,
+        float window_expansion,
+        int max_window_size
+    ) {
+        super(tag, description, ignoreFailure);
+        this.enabled = enabled;
+        if (ratio < 0f || ratio > 1f) {
```

Comment:
Can we make these configuration optional with default value?

Reply:
The optional and default value are in the Factor class.
```
            boolean enabled = ConfigurationUtils.readBooleanProperty(TYPE, tag, config, ENABLE_KEY, DEFAULT_ENABLED);
            Map<String, Object> twoPhaseConfigMap = ConfigurationUtils.readOptionalMap(TYPE, tag, config, PARAMETER_KEY);

            float ratio = DEFAULT_RATIO;
            float window_expansion = DEFAULT_WINDOW_EXPANSION;
            int max_window_size = DEFAULT_MAX_WINDOW_SIZE;
            if (Objects.nonNull(twoPhaseConfigMap)) {
                ratio = ((Number) twoPhaseConfigMap.getOrDefault(RATIO_KEY, ratio)).floatValue();
                window_expansion = ((Number) twoPhaseConfigMap.getOrDefault(EXPANSION_KEY, window_expansion)).floatValue();
                max_window_size = ((Number) twoPhaseConfigMap.getOrDefault(MAX_WINDOW_SIZE_KEY, max_window_size)).intValue();
            }
```